crates/snark-acct-sys/src/verification.rs

```diff
@@ -0,0 +1,194 @@
+        let hash = StrataHasher::hash_leaf(&msg_bytes);
+
+        let cohashes: Vec<[u8; 32]> = msg_proof.raw_proof().cohashes();
+        let proof = MerkleProof::from_cohashes(cohashes, cur_index);
+
+        if !generic_mmr.verify(&proof, &hash) {
+            return Err(AcctError::InvalidMessageProof {
+                account_id,
+                msg_idx: cur_index,
+            });
+        }
+
+        cur_index += 1;
+    }
+    Ok(())
+}
+
+pub(crate) fn verify_update_witness(
+    target: AccountId,
+    snark_state: &impl ISnarkAccountState,
+    update: &SnarkAccountUpdate,
+) -> AcctResult<()> {
+    let vk = snark_state.verification_key();
+    let claim: Vec<u8> = compute_update_claim(snark_state, update.operation());
+    let is_valid = vk
+        .verify_claim_witness(&claim, update.update_proof())
+        .is_ok();
+
+    if !is_valid {
+        return Err(AcctError::InvalidUpdateProof { account_id: target });
+    }
+
+    Ok(())
+}
+
+fn compute_update_claim(
+    snark_state: &impl ISnarkAccountState,
+    operation: &UpdateOperationData,
+) -> Vec<u8> {
+    // Use new state, processed messages, old state, refs and outputs to compute claim
+    let cur_state = ProofState::new(
+        snark_state.inner_state_root(),
+        snark_state.next_inbox_msg_idx(),
+    );
+    let pub_params = UpdateProofPubParams::new(
+        cur_state,
+        operation.new_state(),
+        operation.processed_messages().to_vec(),
+        operation.ledger_refs().clone(),
+        operation.outputs().clone(),
+        operation.extra_data().to_vec(),
+    );
+    pub_params.as_ssz_bytes()
+}
+
+fn verify_update_outputs_safe<S: IStateAccessor>(
+    outputs: &UpdateOutputs,
+    state_accessor: &S,
+    cur_balance: BitcoinAmount,
+) -> AcctResult<()> {
+    let transfers = outputs.transfers();
+    let messages = outputs.messages();
+
+    // Check if receivers exist (skip special/system accounts)
+    for t in transfers {
+        if !t.dest().is_special() && !state_accessor.check_account_exists(t.dest())? {
+            return Err(AcctError::MissingExpectedAccount(t.dest()));
+        }
+    }
+
+    for m in messages {
+        if !m.dest().is_special() && !state_accessor.check_account_exists(m.dest())? {
+            return Err(AcctError::MissingExpectedAccount(m.dest()));
+        }
+    }
+
+    let total_sent = outputs
+        .compute_total_value()
+        .ok_or(AcctError::BitcoinAmountOverflow)?;
+
+    // Check if there is sufficient balance.
+    if total_sent > cur_balance {
+        return Err(AcctError::InsufficientBalance {
+            requested: total_sent,
+            available: cur_balance,
+        });
+    }
+    Ok(())
+}
+
+/// Type safe update that indicates it has been verified.
+#[derive(Clone, Debug)]
+pub struct VerifiedUpdate<'a> {
+    operation: &'a UpdateOperationData,
+}
+
+impl<'a> VerifiedUpdate<'a> {
+    pub fn operation(&self) -> &'a UpdateOperationData {
+        self.operation
+    }
+}
```

Comment:
As I said about this pattern in the older PR, this just makes the code harder to understand and doesn't provide any actual value since it requires extracting the data *out* of the "verified form" when doing anything with it, so it should be removed.  The idea of having this pattern of having a "verified form" *is good* in the right context, but this isn't an effective implementation of it and it makes the impl diverge more from the spec.